<a href="https://colab.research.google.com/github/KaueAbbe/Challenge_Data_Science1_Alura/blob/main/Semana%203/DSModelo_MachineLearning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color = 'red' size = 10 > Modelo de Machine Learning </font>
#### **Autor:** Kaue Hermann Abbehausen
#### **Linkedin:** https://www.linkedin.com/in/kaue-abbehausen-5b1922165/
## <font color = 'green'> Contextualizando</font>
Eu atuo como cientista de dados pela operadora de telecomunicações Alura Voz. Na reunião inicial com as pessoas responsáveis pela área de vendas da empresa, foi explicada a importância de se reduzir a Taxa de Evasão de Clientes, conhecido como **Churn Rate**. Basicamente, o **Churn Rate indica o quanto a empresa perdeu de receita ou clientes em um período de tempo.**

Eu sugeri, como passo inicial, a identificação de clientes que teriam uma maior chance de deixar a empresa. Para isso, expliquei que é interessante investigar algumas características de clientes ou dos planos de clientes para tentar CLASSIFICAR estas pessoas como potenciais candidatas a deixar a empresa ou não.

Assim, solicitei o conjunto de dados para começar a explorar, tratar e modelar a partir de agora. Em seguida, o foco será na otimização de cada um dos modelos com a finalidade de obter o melhor resultado para a tomada de decisão da Alura Voz.

* **Os dados foram retirados de uma API e podem ser encontrado no link abaixo:**

https://github.com/sthemonica/alura-voz/blob/main/Dados/Telco-Customer-Churn.json

* **Os dados foram tratados no arquivo Semana 1 e são encontrados no link abaixo:**

https://github.com/KaueAbbe/Challenge_Data_Science1_Alura  
* Versão Raw

https://raw.githubusercontent.com/KaueAbbe/Challenge_Data_Science1_Alura/main/Telco-Customer-Churn-Corrigido.json


* **A análise exploratório dos dados foi feita na semana 2.E agora é o momento de criar um modelo de machine learning em que é possível prever quais os clientes que podem vir a evadir da empresa AluraVoz**

## **Resultados da Análise exploratória**

Durante a análise exploratória foram encontrados variáveis que apresentam diferença entre os grupos que evadiram e que não evadiram. Para as variáveis do tipo qualitativas (categóricas) temos:

* Idoso,
 Parceiro,
 Dependencia,
 Tipo_contrato,
 Fatura_online,
 Metodo_pagamento,
 Servico_internet,
 Assinatura_tv,
 Assinatura_filme

 Desta lista de variáveis categóricas apenas Idoso, Parceiro, dependencia, Fatura_online, servico_internet, assinatura_tv e assinatura_filme são de duas categorias (sim ou não).

In [ ]:
import pandas as pd


In [ ]:
dados = pd.read_json('https://raw.githubusercontent.com/KaueAbbe/Challenge_Data_Science1_Alura/main/Telco-Customer-Churn-Corrigido.json')
dados[['Evadiu','Idoso',
 'Parceiro',
 'Dependencia',
 'Tipo_contrato',
 'Fatura_online',
 'Metodo_pagamento',
 'Servico_internet',
 'Assinatura_tv',
 'Assinatura_filme']].head(10)

Evadiu Idoso Parceiro Dependencia Tipo_contrato Fatura_online  \
0    Nao   Nao      Sim         Sim         Anual           Sim   
1    Nao   Nao      Nao         Nao     Mes a Mes           Nao   
2    Sim   Nao      Nao         Nao     Mes a Mes           Sim   
3    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
4    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
5    Nao   Nao      Nao         Sim     Mes a Mes           Sim   
6    Nao   Sim      Sim         Nao       Bianual           Sim   
7    Nao   Nao      Sim         Nao       Bianual           Sim   
8    Nao   Sim      Nao         Nao     Mes a Mes           Sim   
9    Nao   Nao      Sim         Sim       Bianual           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  
0              Nao  
1              Sim  
2              Nao  
3              Sim  
4              Nao  
5              Sim  
6              Sim  
7              Nao  
8              Nao  
9              Sim

## 1º Modelo: **Modelo de Classificação**

Modelo leva em consideração os sims e nãos dos dados de variáveis categóricas. AS variáveis que tem mais de uma alternativa de escolha serão transformados em colunas de sim e não. 
Não -> 0
Sim -> 1 

In [ ]:
variaveis_cat = dados[['Evadiu','Idoso',
 'Parceiro',
 'Dependencia',
 'Tipo_contrato',
 'Fatura_online',
 'Metodo_pagamento',
 'Servico_internet',
 'Assinatura_tv',
 'Assinatura_filme']].copy()
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Tipo_contrato Fatura_online  \
0    Nao   Nao      Sim         Sim         Anual           Sim   
1    Nao   Nao      Nao         Nao     Mes a Mes           Nao   
2    Sim   Nao      Nao         Nao     Mes a Mes           Sim   
3    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
4    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
5    Nao   Nao      Nao         Sim     Mes a Mes           Sim   
6    Nao   Sim      Sim         Nao       Bianual           Sim   
7    Nao   Nao      Sim         Nao       Bianual           Sim   
8    Nao   Sim      Nao         Nao     Mes a Mes           Sim   
9    Nao   Nao      Sim         Sim       Bianual           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  
0              Nao  
1              Sim  
2              Nao  
3              Sim  
4              Nao  
5              Sim  
6              Sim  
7              Nao  
8              Nao  
9              Sim

### 1. Alterando Tipo_contrato

In [ ]:
anual = variaveis_cat['Tipo_contrato'].isin(['Anual'])
lista_anual = []
for valor in anual.values:
  if valor == True:
    lista_anual.append(1)
  else:
    lista_anual.append(0)


In [ ]:
variaveis_cat['Contrato_anual'] = lista_anual
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Tipo_contrato Fatura_online  \
0    Nao   Nao      Sim         Sim         Anual           Sim   
1    Nao   Nao      Nao         Nao     Mes a Mes           Nao   
2    Sim   Nao      Nao         Nao     Mes a Mes           Sim   
3    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
4    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
5    Nao   Nao      Nao         Sim     Mes a Mes           Sim   
6    Nao   Sim      Sim         Nao       Bianual           Sim   
7    Nao   Nao      Sim         Nao       Bianual           Sim   
8    Nao   Sim      Nao         Nao     Mes a Mes           Sim   
9    Nao   Nao      Sim         Sim       Bianual           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  
0              Nao               1  
1              Sim               0  
2              Nao               0  
3              Sim               0  
4              Nao               0  
5              Sim               0  
6              Sim               0  
7              Nao               0  
8              Nao               0  
9              Sim               0

In [ ]:
mes = variaveis_cat['Tipo_contrato'].isin(['Mes a Mes'])
lista_mes = []
for valor in mes.values:
  if valor == True:
    lista_mes.append(1)
  else:
    lista_mes.append(0)
variaveis_cat['Contrato_mes'] = lista_mes
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Tipo_contrato Fatura_online  \
0    Nao   Nao      Sim         Sim         Anual           Sim   
1    Nao   Nao      Nao         Nao     Mes a Mes           Nao   
2    Sim   Nao      Nao         Nao     Mes a Mes           Sim   
3    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
4    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
5    Nao   Nao      Nao         Sim     Mes a Mes           Sim   
6    Nao   Sim      Sim         Nao       Bianual           Sim   
7    Nao   Nao      Sim         Nao       Bianual           Sim   
8    Nao   Sim      Nao         Nao     Mes a Mes           Sim   
9    Nao   Nao      Sim         Sim       Bianual           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  
0              Nao               1             0  
1              Sim               0             1  
2              Nao               0             1  
3              Sim               0             1  
4              Nao               0             1  
5              Sim               0             1  
6              Sim               0             0  
7              Nao               0             0  
8              Nao               0             1  
9              Sim               0             0

In [ ]:
bianual = variaveis_cat['Tipo_contrato'].isin(['Bianual'])
lista_bianual = []
for valor in bianual.values:
  if valor == True:
    lista_bianual.append(1)
  else:
    lista_bianual.append(0)
variaveis_cat['Contrato_bianual'] = lista_bianual
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Tipo_contrato Fatura_online  \
0    Nao   Nao      Sim         Sim         Anual           Sim   
1    Nao   Nao      Nao         Nao     Mes a Mes           Nao   
2    Sim   Nao      Nao         Nao     Mes a Mes           Sim   
3    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
4    Sim   Sim      Sim         Nao     Mes a Mes           Sim   
5    Nao   Nao      Nao         Sim     Mes a Mes           Sim   
6    Nao   Sim      Sim         Nao       Bianual           Sim   
7    Nao   Nao      Sim         Nao       Bianual           Sim   
8    Nao   Sim      Nao         Nao     Mes a Mes           Sim   
9    Nao   Nao      Sim         Sim       Bianual           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  
0              Nao               1             0                 0  
1              Sim               0             1                 0  
2              Nao               0             1                 0  
3              Sim               0             1                 0  
4              Nao               0             1                 0  
5              Sim               0             1                 0  
6              Sim               0             0                 1  
7              Nao               0             0                 1  
8              Nao               0             1                 0  
9              Sim               0             0                 1

In [ ]:
variaveis_cat.drop(['Tipo_contrato'], axis = 1, inplace = True)
variaveis_cat.head()

Evadiu Idoso Parceiro Dependencia Fatura_online   Metodo_pagamento  \
0    Nao   Nao      Sim         Sim           Sim   Cheque por Email   
1    Nao   Nao      Nao         Nao           Nao   Cheque por Email   
2    Sim   Nao      Nao         Nao           Sim  Cheque Eletronico   
3    Sim   Sim      Sim         Nao           Sim  Cheque Eletronico   
4    Sim   Sim      Sim         Nao           Sim   Cheque por Email   

  Servico_internet Assinatura_tv Assinatura_filme  Contrato_anual  \
0              DSL           Sim              Nao               1   
1              DSL           Nao              Sim               0   
2     Fibra optica           Nao              Nao               0   
3     Fibra optica           Sim              Sim               0   
4     Fibra optica           Sim              Nao               0   

   Contrato_mes  Contrato_bianual  
0             0                 0  
1             1                 0  
2             1                 0  
3             1                 0  
4             1                 0

### 2. Alterando Metodo_pagamento

In [ ]:
variaveis_cat['Metodo_pagamento'].unique()

array(['Cheque por Email', 'Cheque Eletronico',
       'Cartao Credito (automatico)',
       'Transferencia Bancaria (automatico)'], dtype=object)

In [ ]:
lista_ce = []
for valor in variaveis_cat['Metodo_pagamento'].isin(['Cheque por Email']):
  if valor == True:
    lista_ce.append(1)
  else:
    lista_ce.append(0)
variaveis_cat['Pagamento_Cheque_Email'] = lista_ce
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online  \
0    Nao   Nao      Sim         Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim   
3    Sim   Sim      Sim         Nao           Sim   
4    Sim   Sim      Sim         Nao           Sim   
5    Nao   Nao      Nao         Sim           Sim   
6    Nao   Sim      Sim         Nao           Sim   
7    Nao   Nao      Sim         Nao           Sim   
8    Nao   Sim      Nao         Nao           Sim   
9    Nao   Nao      Sim         Sim           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   
5              Sim               0             1                 0   
6              Sim               0             0                 1   
7              Nao               0             0                 1   
8              Nao               0             1                 0   
9              Sim               0             0                 1   

   Pagamento_Cheque_Email  
0                       1  
1                       1  
2                       0  
3                       0  
4                       1  
5                       0  
6                       0  
7                       0  
8                       0  
9                       1

In [ ]:
lista_cel = []
for valor in variaveis_cat['Metodo_pagamento'].isin(['Cheque Eletronico']):
  if valor == True:
    lista_cel.append(1)
  else:
    lista_cel.append(0)
variaveis_cat['Pagamento_Cheque_Eletrico'] = lista_cel
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online  \
0    Nao   Nao      Sim         Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim   
3    Sim   Sim      Sim         Nao           Sim   
4    Sim   Sim      Sim         Nao           Sim   
5    Nao   Nao      Nao         Sim           Sim   
6    Nao   Sim      Sim         Nao           Sim   
7    Nao   Nao      Sim         Nao           Sim   
8    Nao   Sim      Nao         Nao           Sim   
9    Nao   Nao      Sim         Sim           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   
5              Sim               0             1                 0   
6              Sim               0             0                 1   
7              Nao               0             0                 1   
8              Nao               0             1                 0   
9              Sim               0             0                 1   

   Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  
0                       1                          0  
1                       1                          0  
2                       0                          1  
3                       0                          1  
4                       1                          0  
5                       0                          0  
6                       0                          0  
7                       0                          0  
8                       0                          1  
9                       1                          0

In [ ]:
lista_cartao = []
for valor in variaveis_cat['Metodo_pagamento'].isin(['Cartao Credito (automatico)']):
  if valor == True:
    lista_cartao.append(1)
  else:
    lista_cartao.append(0)
variaveis_cat['Pagamento_Cartao_credito'] = lista_cartao
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online  \
0    Nao   Nao      Sim         Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim   
3    Sim   Sim      Sim         Nao           Sim   
4    Sim   Sim      Sim         Nao           Sim   
5    Nao   Nao      Nao         Sim           Sim   
6    Nao   Sim      Sim         Nao           Sim   
7    Nao   Nao      Sim         Nao           Sim   
8    Nao   Sim      Nao         Nao           Sim   
9    Nao   Nao      Sim         Sim           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   
5              Sim               0             1                 0   
6              Sim               0             0                 1   
7              Nao               0             0                 1   
8              Nao               0             1                 0   
9              Sim               0             0                 1   

   Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  Pagamento_Cartao_credito  
0                       1                          0                         0  
1                       1                          0                         0  
2                       0                          1                         0  
3                       0                          1                         0  
4                       1                          0                         0  
5                       0                          0                         1  
6                       0                          0                         0  
7                       0                          0                         1  
8                       0                          1                         0  
9                       1                          0                         0

In [ ]:
lista_trans = []
for valor in variaveis_cat['Metodo_pagamento'].isin(['Transferencia Bancaria (automatico)']):
  if valor == True:
    lista_trans.append(1)
  else:
    lista_trans.append(0)
variaveis_cat['Pagamento_Transferencia'] = lista_trans
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online  \
0    Nao   Nao      Sim         Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim   
3    Sim   Sim      Sim         Nao           Sim   
4    Sim   Sim      Sim         Nao           Sim   
5    Nao   Nao      Nao         Sim           Sim   
6    Nao   Sim      Sim         Nao           Sim   
7    Nao   Nao      Sim         Nao           Sim   
8    Nao   Sim      Nao         Nao           Sim   
9    Nao   Nao      Sim         Sim           Sim   

                      Metodo_pagamento Servico_internet Assinatura_tv  \
0                     Cheque por Email              DSL           Sim   
1                     Cheque por Email              DSL           Nao   
2                    Cheque Eletronico     Fibra optica           Nao   
3                    Cheque Eletronico     Fibra optica           Sim   
4                     Cheque por Email     Fibra optica           Sim   
5          Cartao Credito (automatico)              DSL           Sim   
6  Transferencia Bancaria (automatico)     Fibra optica           Sim   
7          Cartao Credito (automatico)     Fibra optica           Nao   
8                    Cheque Eletronico              DSL           Nao   
9                     Cheque por Email              DSL           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   
5              Sim               0             1                 0   
6              Sim               0             0                 1   
7              Nao               0             0                 1   
8              Nao               0             1                 0   
9              Sim               0             0                 1   

   Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                       1                          0   
1                       1                          0   
2                       0                          1   
3                       0                          1   
4                       1                          0   
5                       0                          0   
6                       0                          0   
7                       0                          0   
8                       0                          1   
9                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  
0                         0                        0  
1                         0                        0  
2                         0                        0  
3                         0                        0  
4                         0                        0  
5                         1                        0  
6                         0                        1  
7                         1                        0  
8                         0                        0  
9                         0                        0

In [ ]:
variaveis_cat.drop(['Metodo_pagamento'], axis = 1, inplace = True)
variaveis_cat.head()

Evadiu Idoso Parceiro Dependencia Fatura_online Servico_internet  \
0    Nao   Nao      Sim         Sim           Sim              DSL   
1    Nao   Nao      Nao         Nao           Nao              DSL   
2    Sim   Nao      Nao         Nao           Sim     Fibra optica   
3    Sim   Sim      Sim         Nao           Sim     Fibra optica   
4    Sim   Sim      Sim         Nao           Sim     Fibra optica   

  Assinatura_tv Assinatura_filme  Contrato_anual  Contrato_mes  \
0           Sim              Nao               1             0   
1           Nao              Sim               0             1   
2           Nao              Nao               0             1   
3           Sim              Sim               0             1   
4           Sim              Nao               0             1   

   Contrato_bianual  Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                 0                       1                          0   
1                 0                       1                          0   
2                 0                       0                          1   
3                 0                       0                          1   
4                 0                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  
0                         0                        0  
1                         0                        0  
2                         0                        0  
3                         0                        0  
4                         0                        0

### 3. Alterando Servico_internet

In [ ]:
variaveis_cat['Servico_internet'].unique()

array(['DSL', 'Fibra optica', 'Nao'], dtype=object)

In [ ]:
lista_DSL = []
for valor in variaveis_cat['Servico_internet'].isin(['DSL']):
  if valor == True:
    lista_DSL.append(1)
  else:
    lista_DSL.append(0)
variaveis_cat['Internet_DSL'] = lista_DSL
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online Servico_internet  \
0    Nao   Nao      Sim         Sim           Sim              DSL   
1    Nao   Nao      Nao         Nao           Nao              DSL   
2    Sim   Nao      Nao         Nao           Sim     Fibra optica   
3    Sim   Sim      Sim         Nao           Sim     Fibra optica   
4    Sim   Sim      Sim         Nao           Sim     Fibra optica   
5    Nao   Nao      Nao         Sim           Sim              DSL   
6    Nao   Sim      Sim         Nao           Sim     Fibra optica   
7    Nao   Nao      Sim         Nao           Sim     Fibra optica   
8    Nao   Sim      Nao         Nao           Sim              DSL   
9    Nao   Nao      Sim         Sim           Sim              DSL   

  Assinatura_tv Assinatura_filme  Contrato_anual  Contrato_mes  \
0           Sim              Nao               1             0   
1           Nao              Sim               0             1   
2           Nao              Nao               0             1   
3           Sim              Sim               0             1   
4           Sim              Nao               0             1   
5           Sim              Sim               0             1   
6           Sim              Sim               0             0   
7           Nao              Nao               0             0   
8           Nao              Nao               0             1   
9           Sim              Sim               0             0   

   Contrato_bianual  Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                 0                       1                          0   
1                 0                       1                          0   
2                 0                       0                          1   
3                 0                       0                          1   
4                 0                       1                          0   
5                 0                       0                          0   
6                 1                       0                          0   
7                 1                       0                          0   
8                 0                       0                          1   
9                 1                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  
0                         0                        0             1  
1                         0                        0             1  
2                         0                        0             0  
3                         0                        0             0  
4                         0                        0             0  
5                         1                        0             1  
6                         0                        1             0  
7                         1                        0             0  
8                         0                        0             1  
9                         0                        0             1

In [ ]:
lista_fibra = []
for valor in variaveis_cat['Servico_internet'].isin(['Fibra optica']):
  if valor == True:
    lista_fibra.append(1)
  else:
    lista_fibra.append(0)
variaveis_cat['Internet_FibraOptica'] = lista_fibra
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online Servico_internet  \
0    Nao   Nao      Sim         Sim           Sim              DSL   
1    Nao   Nao      Nao         Nao           Nao              DSL   
2    Sim   Nao      Nao         Nao           Sim     Fibra optica   
3    Sim   Sim      Sim         Nao           Sim     Fibra optica   
4    Sim   Sim      Sim         Nao           Sim     Fibra optica   
5    Nao   Nao      Nao         Sim           Sim              DSL   
6    Nao   Sim      Sim         Nao           Sim     Fibra optica   
7    Nao   Nao      Sim         Nao           Sim     Fibra optica   
8    Nao   Sim      Nao         Nao           Sim              DSL   
9    Nao   Nao      Sim         Sim           Sim              DSL   

  Assinatura_tv Assinatura_filme  Contrato_anual  Contrato_mes  \
0           Sim              Nao               1             0   
1           Nao              Sim               0             1   
2           Nao              Nao               0             1   
3           Sim              Sim               0             1   
4           Sim              Nao               0             1   
5           Sim              Sim               0             1   
6           Sim              Sim               0             0   
7           Nao              Nao               0             0   
8           Nao              Nao               0             1   
9           Sim              Sim               0             0   

   Contrato_bianual  Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                 0                       1                          0   
1                 0                       1                          0   
2                 0                       0                          1   
3                 0                       0                          1   
4                 0                       1                          0   
5                 0                       0                          0   
6                 1                       0                          0   
7                 1                       0                          0   
8                 0                       0                          1   
9                 1                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  \
0                         0                        0             1   
1                         0                        0             1   
2                         0                        0             0   
3                         0                        0             0   
4                         0                        0             0   
5                         1                        0             1   
6                         0                        1             0   
7                         1                        0             0   
8                         0                        0             1   
9                         0                        0             1   

   Internet_FibraOptica  
0                     0  
1                     0  
2                     1  
3                     1  
4                     1  
5                     0  
6                     1  
7                     1  
8                     0  
9                     0

In [ ]:
lista_nao = []
for valor in variaveis_cat['Servico_internet'].isin(['Nao']):
  if valor == True:
    lista_nao.append(1)
  else:
    lista_nao.append(0)
variaveis_cat['Internet_SemInternet'] = lista_nao
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online Servico_internet  \
0    Nao   Nao      Sim         Sim           Sim              DSL   
1    Nao   Nao      Nao         Nao           Nao              DSL   
2    Sim   Nao      Nao         Nao           Sim     Fibra optica   
3    Sim   Sim      Sim         Nao           Sim     Fibra optica   
4    Sim   Sim      Sim         Nao           Sim     Fibra optica   
5    Nao   Nao      Nao         Sim           Sim              DSL   
6    Nao   Sim      Sim         Nao           Sim     Fibra optica   
7    Nao   Nao      Sim         Nao           Sim     Fibra optica   
8    Nao   Sim      Nao         Nao           Sim              DSL   
9    Nao   Nao      Sim         Sim           Sim              DSL   

  Assinatura_tv Assinatura_filme  Contrato_anual  Contrato_mes  \
0           Sim              Nao               1             0   
1           Nao              Sim               0             1   
2           Nao              Nao               0             1   
3           Sim              Sim               0             1   
4           Sim              Nao               0             1   
5           Sim              Sim               0             1   
6           Sim              Sim               0             0   
7           Nao              Nao               0             0   
8           Nao              Nao               0             1   
9           Sim              Sim               0             0   

   Contrato_bianual  Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                 0                       1                          0   
1                 0                       1                          0   
2                 0                       0                          1   
3                 0                       0                          1   
4                 0                       1                          0   
5                 0                       0                          0   
6                 1                       0                          0   
7                 1                       0                          0   
8                 0                       0                          1   
9                 1                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  \
0                         0                        0             1   
1                         0                        0             1   
2                         0                        0             0   
3                         0                        0             0   
4                         0                        0             0   
5                         1                        0             1   
6                         0                        1             0   
7                         1                        0             0   
8                         0                        0             1   
9                         0                        0             1   

   Internet_FibraOptica  Internet_SemInternet  
0                     0                     0  
1                     0                     0  
2                     1                     0  
3                     1                     0  
4                     1                     0  
5                     0                     0  
6                     1                     0  
7                     1                     0  
8                     0                     0  
9                     0                     0

In [ ]:
variaveis_cat.drop(['Servico_internet'], axis = 1, inplace = True)
variaveis_cat.head()

Evadiu Idoso Parceiro Dependencia Fatura_online Assinatura_tv  \
0    Nao   Nao      Sim         Sim           Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim           Nao   
3    Sim   Sim      Sim         Nao           Sim           Sim   
4    Sim   Sim      Sim         Nao           Sim           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   

   Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                       1                          0   
1                       1                          0   
2                       0                          1   
3                       0                          1   
4                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  \
0                         0                        0             1   
1                         0                        0             1   
2                         0                        0             0   
3                         0                        0             0   
4                         0                        0             0   

   Internet_FibraOptica  Internet_SemInternet  
0                     0                     0  
1                     0                     0  
2                     1                     0  
3                     1                     0  
4                     1                     0

### 4. Alterando as variáveis Assinatura.
Estas apresentam 3 valores, sendo um deles repetido na variáveis Sem internet

In [ ]:
variaveis_cat['Assinatura_tv'].unique()

array(['Sim', 'Nao', 'Sem servico internet'], dtype=object)

In [ ]:
lista_Assnao = []
for valor in variaveis_cat['Assinatura_tv'].isin(['Sim']):
  if valor == True:
    lista_Assnao.append(1)
  else:
    lista_Assnao.append(0)
variaveis_cat['Tv'] = lista_Assnao
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online Assinatura_tv  \
0    Nao   Nao      Sim         Sim           Sim           Sim   
1    Nao   Nao      Nao         Nao           Nao           Nao   
2    Sim   Nao      Nao         Nao           Sim           Nao   
3    Sim   Sim      Sim         Nao           Sim           Sim   
4    Sim   Sim      Sim         Nao           Sim           Sim   
5    Nao   Nao      Nao         Sim           Sim           Sim   
6    Nao   Sim      Sim         Nao           Sim           Sim   
7    Nao   Nao      Sim         Nao           Sim           Nao   
8    Nao   Sim      Nao         Nao           Sim           Nao   
9    Nao   Nao      Sim         Sim           Sim           Sim   

  Assinatura_filme  Contrato_anual  Contrato_mes  Contrato_bianual  \
0              Nao               1             0                 0   
1              Sim               0             1                 0   
2              Nao               0             1                 0   
3              Sim               0             1                 0   
4              Nao               0             1                 0   
5              Sim               0             1                 0   
6              Sim               0             0                 1   
7              Nao               0             0                 1   
8              Nao               0             1                 0   
9              Sim               0             0                 1   

   Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                       1                          0   
1                       1                          0   
2                       0                          1   
3                       0                          1   
4                       1                          0   
5                       0                          0   
6                       0                          0   
7                       0                          0   
8                       0                          1   
9                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  \
0                         0                        0             1   
1                         0                        0             1   
2                         0                        0             0   
3                         0                        0             0   
4                         0                        0             0   
5                         1                        0             1   
6                         0                        1             0   
7                         1                        0             0   
8                         0                        0             1   
9                         0                        0             1   

   Internet_FibraOptica  Internet_SemInternet  Tv  
0                     0                     0   1  
1                     0                     0   0  
2                     1                     0   0  
3                     1                     0   1  
4                     1                     0   1  
5                     0                     0   1  
6                     1                     0   1  
7                     1                     0   0  
8                     0                     0   0  
9                     0                     0   1

In [ ]:
variaveis_cat.drop(['Assinatura_tv'], axis = 1, inplace = True)
variaveis_cat.head()

Evadiu Idoso Parceiro Dependencia Fatura_online Assinatura_filme  \
0    Nao   Nao      Sim         Sim           Sim              Nao   
1    Nao   Nao      Nao         Nao           Nao              Sim   
2    Sim   Nao      Nao         Nao           Sim              Nao   
3    Sim   Sim      Sim         Nao           Sim              Sim   
4    Sim   Sim      Sim         Nao           Sim              Nao   

   Contrato_anual  Contrato_mes  Contrato_bianual  Pagamento_Cheque_Email  \
0               1             0                 0                       1   
1               0             1                 0                       1   
2               0             1                 0                       0   
3               0             1                 0                       0   
4               0             1                 0                       1   

   Pagamento_Cheque_Eletrico  Pagamento_Cartao_credito  \
0                          0                         0   
1                          0                         0   
2                          1                         0   
3                          1                         0   
4                          0                         0   

   Pagamento_Transferencia  Internet_DSL  Internet_FibraOptica  \
0                        0             1                     0   
1                        0             1                     0   
2                        0             0                     1   
3                        0             0                     1   
4                        0             0                     1   

   Internet_SemInternet  Tv  
0                     0   1  
1                     0   0  
2                     0   0  
3                     0   1  
4                     0   1

In [ ]:
lista_Ass = []
for valor in variaveis_cat['Assinatura_filme'].isin(['Sim']):
  if valor == True:
    lista_Ass.append(1)
  else:
    lista_Ass.append(0)
variaveis_cat['Filme_sim'] = lista_Ass
variaveis_cat.head(10)

Evadiu Idoso Parceiro Dependencia Fatura_online Assinatura_filme  \
0    Nao   Nao      Sim         Sim           Sim              Nao   
1    Nao   Nao      Nao         Nao           Nao              Sim   
2    Sim   Nao      Nao         Nao           Sim              Nao   
3    Sim   Sim      Sim         Nao           Sim              Sim   
4    Sim   Sim      Sim         Nao           Sim              Nao   
5    Nao   Nao      Nao         Sim           Sim              Sim   
6    Nao   Sim      Sim         Nao           Sim              Sim   
7    Nao   Nao      Sim         Nao           Sim              Nao   
8    Nao   Sim      Nao         Nao           Sim              Nao   
9    Nao   Nao      Sim         Sim           Sim              Sim   

   Contrato_anual  Contrato_mes  Contrato_bianual  Pagamento_Cheque_Email  \
0               1             0                 0                       1   
1               0             1                 0                       1   
2               0             1                 0                       0   
3               0             1                 0                       0   
4               0             1                 0                       1   
5               0             1                 0                       0   
6               0             0                 1                       0   
7               0             0                 1                       0   
8               0             1                 0                       0   
9               0             0                 1                       1   

   Pagamento_Cheque_Eletrico  Pagamento_Cartao_credito  \
0                          0                         0   
1                          0                         0   
2                          1                         0   
3                          1                         0   
4                          0                         0   
5                          0                         1   
6                          0                         0   
7                          0                         1   
8                          1                         0   
9                          0                         0   

   Pagamento_Transferencia  Internet_DSL  Internet_FibraOptica  \
0                        0             1                     0   
1                        0             1                     0   
2                        0             0                     1   
3                        0             0                     1   
4                        0             0                     1   
5                        0             1                     0   
6                        1             0                     1   
7                        0             0                     1   
8                        0             1                     0   
9                        0             1                     0   

   Internet_SemInternet  Tv  Filme_sim  
0                     0   1          0  
1                     0   0          1  
2                     0   0          0  
3                     0   1          1  
4                     0   1          0  
5                     0   1          1  
6                     0   1          1  
7                     0   0          0  
8                     0   0          0  
9                     0   1          1

In [ ]:
variaveis_cat.drop(['Assinatura_filme'], axis = 1, inplace = True)
variaveis_cat.head()

Evadiu Idoso Parceiro Dependencia Fatura_online  Contrato_anual  \
0    Nao   Nao      Sim         Sim           Sim               1   
1    Nao   Nao      Nao         Nao           Nao               0   
2    Sim   Nao      Nao         Nao           Sim               0   
3    Sim   Sim      Sim         Nao           Sim               0   
4    Sim   Sim      Sim         Nao           Sim               0   

   Contrato_mes  Contrato_bianual  Pagamento_Cheque_Email  \
0             0                 0                       1   
1             1                 0                       1   
2             1                 0                       0   
3             1                 0                       0   
4             1                 0                       1   

   Pagamento_Cheque_Eletrico  Pagamento_Cartao_credito  \
0                          0                         0   
1                          0                         0   
2                          1                         0   
3                          1                         0   
4                          0                         0   

   Pagamento_Transferencia  Internet_DSL  Internet_FibraOptica  \
0                        0             1                     0   
1                        0             1                     0   
2                        0             0                     1   
3                        0             0                     1   
4                        0             0                     1   

   Internet_SemInternet  Tv  Filme_sim  
0                     0   1          0  
1                     0   0          1  
2                     0   0          0  
3                     0   1          1  
4                     0   1          0

### 5. Alterando sims e não de outras variáveis

In [ ]:
traducao ={
    'Sim':1,
    'Nao':0
}

In [ ]:
for coluna in ['Evadiu','Idoso','Parceiro','Dependencia','Fatura_online']:
  variaveis_cat[coluna] = variaveis_cat[coluna].map(traducao)
variaveis_cat.head(10)

Evadiu  Idoso  Parceiro  Dependencia  Fatura_online  Contrato_anual  \
0       0      0         1            1              1               1   
1       0      0         0            0              0               0   
2       1      0         0            0              1               0   
3       1      1         1            0              1               0   
4       1      1         1            0              1               0   
5       0      0         0            1              1               0   
6       0      1         1            0              1               0   
7       0      0         1            0              1               0   
8       0      1         0            0              1               0   
9       0      0         1            1              1               0   

   Contrato_mes  Contrato_bianual  Pagamento_Cheque_Email  \
0             0                 0                       1   
1             1                 0                       1   
2             1                 0                       0   
3             1                 0                       0   
4             1                 0                       1   
5             1                 0                       0   
6             0                 1                       0   
7             0                 1                       0   
8             1                 0                       0   
9             0                 1                       1   

   Pagamento_Cheque_Eletrico  Pagamento_Cartao_credito  \
0                          0                         0   
1                          0                         0   
2                          1                         0   
3                          1                         0   
4                          0                         0   
5                          0                         1   
6                          0                         0   
7                          0                         1   
8                          1                         0   
9                          0                         0   

   Pagamento_Transferencia  Internet_DSL  Internet_FibraOptica  \
0                        0             1                     0   
1                        0             1                     0   
2                        0             0                     1   
3                        0             0                     1   
4                        0             0                     1   
5                        0             1                     0   
6                        1             0                     1   
7                        0             0                     1   
8                        0             1                     0   
9                        0             1                     0   

   Internet_SemInternet  Tv  Filme_sim  
0                     0   1          0  
1                     0   0          1  
2                     0   0          0  
3                     0   1          1  
4                     0   1          0  
5                     0   1          1  
6                     0   1          1  
7                     0   0          0  
8                     0   0          0  
9                     0   1          1

### 6. Estudando os dados

In [ ]:
variaveis_cat.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Evadiu                     7043 non-null   int64
 1   Idoso                      7043 non-null   int64
 2   Parceiro                   7043 non-null   int64
 3   Dependencia                7043 non-null   int64
 4   Fatura_online              7043 non-null   int64
 5   Contrato_anual             7043 non-null   int64
 6   Contrato_mes               7043 non-null   int64
 7   Contrato_bianual           7043 non-null   int64
 8   Pagamento_Cheque_Email     7043 non-null   int64
 9   Pagamento_Cheque_Eletrico  7043 non-null   int64
 10  Pagamento_Cartao_credito   7043 non-null   int64
 11  Pagamento_Transferencia    7043 non-null   int64
 12  Internet_DSL               7043 non-null   int64
 13  Internet_FibraOptica       7043 non-null   int64
 14  Internet_SemInternet    

In [ ]:
nao_evasores = round(variaveis_cat['Evadiu'].value_counts(normalize = True)[0],1)
evasores = round(variaveis_cat['Evadiu'].value_counts(normalize = True)[1],1)
print(nao_evasores, evasores)

0.7 0.3


In [ ]:
print(f'Temos {len(variaveis_cat)} dados. Sendo que {round(len(variaveis_cat) * nao_evasores,0)} são Não Evasores, {round(len(variaveis_cat) * evasores,0)} são Evasores')

Temos 7043 dados. Sendo que 4930.0 são Não Evasores, 2113.0 são Evasores


Para cada 1 Evasor existe pouco mais que 2 Permanecentes

### 7. **Criando o Modelo**

In [ ]:
variaveis_cat.columns

Index(['Evadiu', 'Idoso', 'Parceiro', 'Dependencia', 'Fatura_online',
       'Contrato_anual', 'Contrato_mes', 'Contrato_bianual',
       'Pagamento_Cheque_Email', 'Pagamento_Cheque_Eletrico',
       'Pagamento_Cartao_credito', 'Pagamento_Transferencia', 'Internet_DSL',
       'Internet_FibraOptica', 'Internet_SemInternet', 'Tv', 'Filme_sim'],
      dtype='object')

In [ ]:
x = variaveis_cat[['Idoso', 'Parceiro', 'Dependencia', 'Fatura_online',
       'Contrato_anual', 'Contrato_mes', 'Contrato_bianual',
       'Pagamento_Cheque_Email', 'Pagamento_Cheque_Eletrico',
       'Pagamento_Cartao_credito', 'Pagamento_Transferencia', 'Internet_DSL',
       'Internet_FibraOptica', 'Internet_SemInternet', 'Tv', 'Filme_sim']]
x.head()

Idoso  Parceiro  Dependencia  Fatura_online  Contrato_anual  Contrato_mes  \
0      0         1            1              1               1             0   
1      0         0            0              0               0             1   
2      0         0            0              1               0             1   
3      1         1            0              1               0             1   
4      1         1            0              1               0             1   

   Contrato_bianual  Pagamento_Cheque_Email  Pagamento_Cheque_Eletrico  \
0                 0                       1                          0   
1                 0                       1                          0   
2                 0                       0                          1   
3                 0                       0                          1   
4                 0                       1                          0   

   Pagamento_Cartao_credito  Pagamento_Transferencia  Internet_DSL  \
0                         0                        0             1   
1                         0                        0             1   
2                         0                        0             0   
3                         0                        0             0   
4                         0                        0             0   

   Internet_FibraOptica  Internet_SemInternet  Tv  Filme_sim  
0                     0                     0   1          0  
1                     0                     0   0          1  
2                     1                     0   0          0  
3                     1                     0   1          1  
4                     1                     0   1          0

In [ ]:
y = variaveis_cat['Evadiu']
y.head()

0    0
1    0
2    1
3    1
4    1
Name: Evadiu, dtype: int64

In [ ]:
from sklearn.svm import LinearSVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

seed = np.random.seed = 10

treino_x, teste_x, treino_y, teste_y = train_test_split(x, y, test_size = 0.2,random_state = seed, stratify = y)
print(treino_x.shape)
print(teste_x.shape)

modelo = LinearSVC()
modelo.fit(treino_x,treino_y)
previsoes = modelo.predict(teste_x)

acuracia = accuracy_score(teste_y,previsoes)*100
print('A acuracia é de %.2f%%' %acuracia)

(5634, 16)
(1409, 16)
A acuracia é de 78.07%


In [ ]:
from sklearn.dummy import DummyClassifier
dummy = DummyClassifier()
dummy.fit(treino_x, treino_y)
previsoesDummy = dummy.predict(teste_x)
acuracia = accuracy_score(teste_y,previsoesDummy)*100
print('A acuracia é de %.2f%%' %acuracia)


A acuracia é de 73.46%


* Com uma acurácia de 73,46% do Dummy e 78.07% do modelo é interessante repensar o modelo ou criar um novo para aumentar a diferença entre esses dois.
* Neste modelo somente foi usado as variáveis categóricas. Deve-se implementar as demais variáveis.

## 2º Modelo: Em Criação